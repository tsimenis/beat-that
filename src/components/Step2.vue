<template>
  <div :class="$style.step2">
    <h2 :class="$style.title">First tell me your name</h2>
    <p :class="$style.subtitle">Just so I know who I am cheering for!</p>
    <form @submit.prevent="handleSubmit">
      <p v-show="hasError" :class="$style.error">Please fill in your name</p>
      <div :class="$style['input-container']">
        <input
          v-model="playerName"
          :class="$style.input"
          type="text"
          placeholder="Name"
        />
      </div>
      <u-button>
        Next
      </u-button>
    </form>
  </div>
</template>

<script>
  export default {
    data () {
      return {
        playerName: '',
        submitAttempt: false
      }
    },
    computed: {
      hasError () {
        return this.submitAttempt && this.playerName === ''
      }
    },
    methods: {
      handleSubmit () {
        this.submitAttempt = true
        if (this.hasError) return false
        this.$store.commit('SET_NAME', this.playerName)
        this.$store.commit('SET_STEP', 3)
      }
    }
  }
</script>

<style lang="scss" module>

  .title {
    font-size: 2rem;
    margin-bottom: 1rem;
  }

  .subtitle {
    margin-bottom: 4rem;
  }

  .input-container {
    max-width: 75%;
    margin: 0 auto;
  }

  .input {
    width: 100%;
    height: 4rem;
    padding: 0 2rem;
    border-radius: $rounded;
    border: 0;
    outline: none;
    margin-bottom: 2rem;
  }

</style>
