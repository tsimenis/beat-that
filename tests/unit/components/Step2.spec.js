import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import idObj from 'identity-obj-proxy'
import Step2 from '@/components/Step2.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Step2', () => {
  let build
  let store
  let mutations

  beforeEach(() => {
    build = (context = {}) => {
      mutations = {
        SET_NAME: jest.fn(),
        SET_STEP: jest.fn()
      }
      store = new Vuex.Store({
        mutations
      })
      return shallowMount(Step2, {
        ...context,
        store,
        localVue,
        stubs: ['u-button'],
        mocks: {
          $style: idObj
        }
      })
    }
  })

  it('it renders correctly', () => {
    const wrapper = build()
    expect(wrapper.element).toMatchSnapshot()
  })
  it('it displays validation error when form is submitted with empty name', async () => {
    const wrapper = build()
    const error = wrapper.find('.error')

    expect(error.isVisible()).toBe(false)

    const form = wrapper.find('form')
    await form.trigger('submit')

    expect(error.isVisible()).toBe(true)
  })
  it('hides validation error when name is updated after form submit', async () => {
    const wrapper = build()
    const input = wrapper.find('input')
    const error = wrapper.find('.error')
    const form = wrapper.get('form')

    await form.trigger('submit')
    expect(error.isVisible()).toBe(true)

    input.element.value = 'George'
    await input.trigger('input')
    await input.trigger('blur')

    expect(error.isVisible()).toBe(false)
  })
  it('submits the form and commits mutations with correct payload', async () => {
    const wrapper = build()
    const input = wrapper.find('input')
    const form = wrapper.get('form')

    input.element.value = 'George'
    await input.trigger('input')
    await form.trigger('submit')

    expect(mutations.SET_NAME).toHaveBeenCalledWith({}, 'George')
    expect(mutations.SET_STEP).toHaveBeenCalledWith({}, 3)
  })
})
