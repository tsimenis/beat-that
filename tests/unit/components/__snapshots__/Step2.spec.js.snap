// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step2 it renders correctly 1`] = `
<div
  class="step2"
>
  <h2
    class="title"
  >
    First tell me your name
  </h2>
   
  <p
    class="subtitle"
  >
    Just so I know who I am cheering for!
  </p>
   
  <form>
    <p
      class="error"
      style="display: none;"
    >
      Please fill in your name
    </p>
     
    <div
      class="input-container"
    >
      <input
        class="input"
        placeholder="Name"
        type="text"
      />
    </div>
     
    <u-button-stub
      class="form-button"
    >
      
      Next
    
    </u-button-stub>
  </form>
</div>
`;
